lib/google/gax/config.rb

```diff
@@ -0,0 +1,127 @@
+#     * Redistributions in binary form must reproduce the above
+# copyright notice, this list of conditions and the following disclaimer
+# in the documentation and/or other materials provided with the
+# distribution.
+#     * Neither the name of Google LLC nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+module Google
+  module Gax
+    ##
+    # Config is a simple DSL for creating Configuration classes.
+    #
+    # @example
+    #   require "google/gax/config"
+    #
+    #   class SampleConfig
+    #     extend Google::Gax::Config
+    #
+    #     config_attr :name,   nil,         String, nil
+    #     config_attr :active, true,        true, false
+    #     config_attr :count,  nil,         Numeric, nil
+    #     config_attr :env,    :production, String, Symbol
+    #
+    #     def initialize parent_config = nil
+    #       @parent_config = parent_config unless parent_config.nil?
+    #       yield self if block_given?
+    #     end
+    #   end
+    #
+    #   config = SampleConfig.new
+    #
+    #   config.name             #=> nil
+    #   config.name = "thor"    #=> "thor"
+    #   config.name             #=> "thor"
+    #   config.name = :thor     # ArgumentError
+    #
+    module Config
+      ##
+      # Add configuration attribute methods to the configuratin class.
+      #
+      # @param [String, Symbol] key The name of the option
+      # @param [Object, nil] initial Initial value (nil is allowed)
+      # @param [Hash] opts Validation options
+      #
+      def config_attr name, default, *valid_values, &validator
+        name = String(name).to_sym
+        raise "cannot create field named parent_config" if name == :parent_config
+        raise "method #{name} already exists" if method_defined? name
+
+        raise "validation must be provided" if validator.nil? && valid_values.empty?
+        validator ||= ->(value) { valid_values.any? { |v| v === value } }
+
+        name_ivar = "@#{name}".to_sym
+
+        if default.nil?
+          # create getter
+          define_method name do
+            value = instance_variable_get name_ivar
```

Comment:
I think this will still generate warnings until someone sets the instance variable. Is there a way to make sure the initialize method initializes all instance variables (maybe including `@parent_config`) automatically?

Reply:
Dang it, I just checked and you are correct. I thought this was not raising a warning. I'll refactor and add a `instance_variable_defined?` conditional.

Reply:
The warnings have been resolved.